pig/src/main/java/com/netflix/iceberg/pig/IcebergInputFormat.java

```diff
@@ -0,0 +1,260 @@
+      return new String[0];
+    }
+
+    @Override
+    public void write(DataOutput out) throws IOException {
+      byte[] data = SerializationUtils.serialize(this.task);
+      out.writeInt(data.length);
+      out.write(data);
+    }
+
+    @Override
+    public void readFields(DataInput in) throws IOException {
+      byte[] data = new byte[in.readInt()];
+      in.readFully(data);
+
+      this.task = (CombinedScanTask) SerializationUtils.deserialize(data);
+    }
+  }
+
+  public class IcebergRecordReader<T> extends RecordReader<Void, T> {
+    private IcebergSplit split;
+    private TaskAttemptContext context;
+    private ReadSupport<T> readSupport;
+
+    private Iterator<FileScanTask> tasks;
+    private FileScanTask currentTask;
+
+    private Iterator<T> recordIterator;
+    private T currentRecord;
+
+    @Override
+    public void initialize(InputSplit split, TaskAttemptContext context) throws IOException, InterruptedException {
+      this.split = (IcebergSplit) split;
+      this.context = context;
+
+      CombinedScanTask ctask = ((IcebergSplit) split).task;
+      tasks = ctask.files().iterator();
+
+      advance();
+    }
+
+    @SuppressWarnings("unchecked")
+    private boolean advance() throws IOException {
+      if (!tasks.hasNext()) {
+        return false;
+      }
+      currentTask = tasks.next();
+
+      Schema tableSchema = (Schema) ObjectSerializer.deserialize(context.getConfiguration().get(ICEBERG_SCHEMA));
+      List<String> projectedFields = (List<String>) ObjectSerializer.deserialize(context.getConfiguration().get(ICEBERG_PROJECTED_FIELDS));
+
+      Schema projectedSchema = projectedFields != null ? project(tableSchema, projectedFields) : tableSchema;
+
+      PartitionSpec spec = currentTask.asFileScanTask().spec();
+      DataFile file = currentTask.file();
+      InputFile inputFile = HadoopInputFile.fromLocation(file.path(), context.getConfiguration());
+
+      Set<Integer> idColumns = spec.identitySourceIds();
+
+      // schema needed for the projection and filtering
+      boolean hasJoinedPartitionColumns = !idColumns.isEmpty();
+
+      switch (file.format()) {
+        case PARQUET:
+          Map<Integer, Object> partitionValueMap = Maps.newHashMap();
+
+          if (hasJoinedPartitionColumns) {
+
+            Schema readSchema = TypeUtil.selectNot(projectedSchema, idColumns);
+            Schema partitionSchema = TypeUtil.select(tableSchema, idColumns);
+            Schema projectedPartitionSchema = TypeUtil.select(projectedSchema, idColumns);
+
+            for (Types.NestedField field : projectedPartitionSchema.columns()) {
+              int tupleIndex = projectedSchema.columns().indexOf(field);
+              int partitionIndex = partitionSchema.columns().indexOf(field);
+
+              partitionValueMap.put(tupleIndex, file.partition().get(partitionIndex, Object.class));
+            }
+
+            recordIterator = (Iterator<T>) Parquet.read(inputFile)
```

Comment:
Looks like this leaks open files. That's not terrible since it is MR and the process will exit, but there's still a chance of combining a bunch of small files into one split and exhausting the connection pool to S3.

An easy fix is to extend [`CloseableGroup`](https://github.com/Netflix/iceberg/blob/master/api/src/main/java/com/netflix/iceberg/io/CloseableGroup.java) and add each `Iterable` using `addCloseable`.

Reply:
Hmm, this seems like something we should be looking into on the record iterator to be self closing.  I guess there could be cases where the consumer doesn't read all the records, but I don't think that's the case for pig.  Regardless, I'll update to use a CloseableGroup as that will cover both cases.